<a href="https://colab.research.google.com/github/KpihX/rocket-durability-ai/blob/main/Data_Cleaning25.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <font color=#023F7C> **Data cleaning and exploration** </font>

<font color=#023F7C>**Hi! PARIS DataBootcamp 2023 🚀**</font> <br>


<img src = https://www.hi-paris.fr/wp-content/uploads/2020/09/logo-hi-paris-retina.png width = "300" height = "200" >

**What is Data Cleaning ?**<br>
Data cleaning is a crucial step in the data analysis and machine learning process, as the quality of the insights and models generated heavily relies on the accuracy and reliability of the underlying data. Raw data often contains **errors**, **inconsistencies**, **missing values**, and **outliers** that can distort results or lead to faulty conclusions. Data cleaning involves identifying and rectifying these issues, ensuring the dataset is trustworthy and suitable for analysis.

Python provides a robust ecosystem of libraries and tools for data cleaning tasks. <br>
Python's versatility in data cleaning contributes significantly to producing accurate analyses and reliable machine learning models.
- The `Pandas`  library offers functions to handle missing data through imputation or removal, detect and remove duplicates, and transform data types.
- The `NumPy` library can assist in dealing with outliers by providing statistical methods for outlier detection and filtering.
- Additionally, visualization libraries like Matplotlib and Seaborn can help visually identify anomalies.


**Before you start to working on this notebook ⚠️**: <br>
Please download/copy this notebook from `hfactory_magic_folders\course` and drop it into your own directory `my_work` on HFactory. <br>
If you don't, you won't be able to save the modifications you've made on this notebook.

**Need help ?** <br>
You can go to the Introduction and Intermediate python notebooks to learn how to use the `pandas` library. <br>

## **1. Import libraries and dataset**
First, let's import Python libraries.

In [43]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as plt
import os

from google.colab import drive

pd.set_option('display.max_columns', None) #Show all columns

Then, let's import the dataset `dataset_train.csv` using pandas.


In [44]:
drive.mount('/content/drive')
os.chdir("/content/drive/MyDrive/Hi_AI_2025/Project")
os.listdir()

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


['dataset.csv', 'dataset_clean.csv', 'Images', 'Data_Cleaning25.ipynb']

In [45]:
dataset = pd.read_csv("dataset.csv")

## **2. Data discovery**

**Question 1**: <br>
**Display the dataset's head and tail.**

In [23]:
dataset.head(10)

engine_unit_number  time_cycles  setting_1  setting_2  setting_3  \
0                   1            1    -0.0007    -0.0004      100.0   
1                   1            2     0.0019    -0.0003      100.0   
2                   1            3    -0.0043     0.0003      100.0   
3                   1            4     0.0007     0.0000      100.0   
4                   1            5    -0.0019    -0.0002      100.0   
5                   1            6    -0.0043    -0.0001      100.0   
6                   1            7     0.0010     0.0001      100.0   
7                   1            8    -0.0034     0.0003      100.0   
8                   1            9     0.0008     0.0001      100.0   
9                   1           10    -0.0033     0.0001      100.0   

   Fan inlet temperature (°C)  LPC outlet temperature (°C)  \
0                        15.0                    83.416667   
1                        15.0                    83.600000   
2                        15.0                    83.711111   
3                        15.0                    83.711111   
4                        15.0                    83.722222   
5                        15.0                    83.572222   
6                        15.0                    83.783333   
7                        15.0                    83.827778   
8                        15.0                    83.583333   
9                        15.0                    83.355556   

  HPC outlet temperature (°C)  LPT outlet temperature (°C)  \
0                      Normal                   504.961111   
1                      Normal                   506.372222   
2                      Normal                   506.961111   
3                    Very Low                   505.666667   
4                         NaN                   508.083333   
5                         Low                   503.722222   
6                      Normal                   503.388889   
7                         NaN                   505.166667   
8                      Normal                   501.738889   
9                         NaN                   504.883333   

   Fan inlet pressure (bar)  Bypass duct pressure (bar)  \
0                  1.008014                    1.489957   
1                  1.008014                    1.489957   
2                  1.008014                    1.489957   
3                  1.008014                    1.489957   
4                  1.008014                    1.489957   
5                  1.008014                    1.489957   
6                  1.008014                    1.489957   
7                  1.008014                    1.489957   
8                  1.008014                    1.489957   
9                  1.008014                    1.489957   

   HPC outlet pressure (bar)  Physical fan speed (rpm)  \
0                  38.221777                   2388.06   
1                  38.179719                   2388.04   
2                  38.214882                   2388.08   
3                  38.227982                   2388.11   
4                  38.196955                   2388.06   
5                  38.243150                   2388.02   
6                  38.220398                   2388.02   
7                  38.186613                   2388.00   
8                  38.175582                   2388.05   
9                  38.168687                   2388.05   

   Physical core speed (rpm)  Engine pressure ratio (P50/P2) (dimensionless)  \
0                    9046.19                                             1.3   
1                    9044.07                                             1.3   
2                    9052.94                                             1.3   
3                    9049.48                                             1.3   
4                    9055.15                                             1.3   
5                    9049.68                                             1.3   
6     

In [25]:
dataset.tail(10)

engine_unit_number  time_cycles  setting_1  setting_2  setting_3  \
20621                 100          191    -0.0005    -0.0000      100.0   
20622                 100          192    -0.0009     0.0001      100.0   
20623                 100          193    -0.0001     0.0002      100.0   
20624                 100          194    -0.0011     0.0003      100.0   
20625                 100          195    -0.0002    -0.0001      100.0   
20626                 100          196    -0.0004    -0.0003      100.0   
20627                 100          197    -0.0016    -0.0005      100.0   
20628                 100          198     0.0004     0.0000      100.0   
20629                 100          199    -0.0011     0.0003      100.0   
20630                 100          200    -0.0032    -0.0005      100.0   

       Fan inlet temperature (°C)  LPC outlet temperature (°C)  \
20621                        15.0                    84.455556   
20622                        15.0                    84.366667   
20623                        15.0                    84.122222   
20624                        15.0                    84.472222   
20625                        15.0                    84.300000   
20626                        15.0                    84.344444   
20627                        15.0                    84.372222   
20628                        15.0                    84.305556   
20629                        15.0                    84.200000   
20630                        15.0                    84.544444   

      HPC outlet temperature (°C)  LPT outlet temperature (°C)  \
20621              Extremely High                   519.733333   
20622                         NaN                   515.450000   
20623                   Very High                   520.700000   
20624                   Very High                   519.855556   
20625                   Very High                   522.350000   
20626                   Very High                   520.533333   
20627              Extremely High                   523.283333   
20628                   Very High                   520.283333   
20629              Extremely High                   519.366667   
20630                   Very High                   522.483333   

       Fan inlet pressure (bar)  Bypass duct pressure (bar)  \
20621                  1.008014                    1.489957   
20622                  1.008014                    1.489957   
20623                  1.008014                    1.489957   
20624                  1.008014                    1.489957   
20625                  1.008014                    1.489957   
20626                  1.008014                    1.489957   
20627                  1.008014                    1.489957   
20628                  1.008014                    1.489957   
20629                  1.008014                    1.489957   
20630                  1.008014                    1.489957   

       HPC outlet pressure (bar)  Physical fan speed (rpm)  \
20621                  38.043892                   2388.26   
20622                  37.999765                   2388.17   
20623                  38.061818                   2388.29   
20624                  38.058371                   2388.23   
20625                  38.019071                   2388.23   
20626                  38.019760                   2388.19   
20627                  37.980460                   2388.23   
20628                  37.985976                   2388.24   
20629                  37.968049                   2388.25   
20630                  37.975634                   2388.26   

       Physical core speed (rpm)  \
20621                    9068.90   
20622                    9060.45   
20623                    9067.57   
20624                    9068.85   
20625                    9069.69   
20626                    9065.52   
20627                    9065.11   
20628                    9065.90   
20629                    9

**Question 2**: <br> **Use the pandas function `.info()` to get general information on the dataset.**<br>

In [26]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20631 entries, 0 to 20630
Data columns (total 28 columns):
 #   Column                                             Non-Null Count  Dtype  
---  ------                                             --------------  -----  
 0   engine_unit_number                                 20631 non-null  int64  
 1   time_cycles                                        20631 non-null  int64  
 2   setting_1                                          20631 non-null  float64
 3   setting_2                                          20631 non-null  float64
 4   setting_3                                          20631 non-null  float64
 5   Fan inlet temperature (°C)                         20631 non-null  float64
 6   LPC outlet temperature (°C)                        20631 non-null  float64
 7   HPC outlet temperature (°C)                        18765 non-null  object 
 8   LPT outlet temperature (°C)                        20631 non-null  float64
 9   Fan in

**What can you say about the loaded dataset ?**


- 3 columns have some nan values
- 3 data types : int64, float64, object
- 28 columns, 20631 lines

**Question 3**:  <br> **Print all the columns/variables of the dataset.**

In [27]:
dataset.columns

Index(['engine_unit_number', 'time_cycles', 'setting_1', 'setting_2',
       'setting_3', 'Fan inlet temperature (°C)',
       'LPC outlet temperature (°C)', 'HPC outlet temperature (°C)',
       'LPT outlet temperature (°C)', 'Fan inlet pressure (bar)',
       'Bypass duct pressure (bar)', 'HPC outlet pressure (bar)',
       'Physical fan speed (rpm)', 'Physical core speed (rpm)',
       'Engine pressure ratio (P50/P2) (dimensionless)',
       'HPC outlet static pressure (bar)', 'Fuel flow / Ps30 (kg/s/bar)',
       'Corrected fan speed (rpm)', 'Corrected core speed (rpm)',
       'Bypass ratio (dimensionless)',
       'Burner fuel-air ratio (mass ratio, dimensionless)',
       'Bleed enthalpy (kJ/kg)', 'Demanded fan speed (rpm)',
       'Demanded corrected fan speed (rpm)', 'HPT coolant bleed flow (kg/s)',
       'LPT coolant bleed flow (kg/s)', 'RUL', 'RUL_class'],
      dtype='object')

## **3. Analyze the dataframe's dtypes**
**Question 4**: <br>
**Create 3 lists, each containing columns names with an int, float and object type.**
- List 1: Columns with an `int64` type
- List 2: Columns with a `float64` type
- List 3: Columns with an `object` type.

*Note: You can use pandas' `.select_dtypes()` function to get columns with a specific dtype.* <br>
*Create a list from a Pandas Dataframe/series with `.to_list()`*

In [28]:
int_columns = dataset.select_dtypes(include=['int64']).columns.to_list()
float_columns = dataset.select_dtypes(include=['float64']).columns.to_list()
object_columns = dataset.select_dtypes(include=['object']).columns.to_list()

print(int_columns)
print(float_columns)
print(object_columns)

['engine_unit_number', 'time_cycles', 'Demanded fan speed (rpm)', 'RUL', 'RUL_class']
['setting_1', 'setting_2', 'setting_3', 'Fan inlet temperature (°C)', 'LPC outlet temperature (°C)', 'LPT outlet temperature (°C)', 'Fan inlet pressure (bar)', 'Bypass duct pressure (bar)', 'HPC outlet pressure (bar)', 'Physical fan speed (rpm)', 'Physical core speed (rpm)', 'Engine pressure ratio (P50/P2) (dimensionless)', 'HPC outlet static pressure (bar)', 'Fuel flow / Ps30 (kg/s/bar)', 'Corrected fan speed (rpm)', 'Corrected core speed (rpm)', 'Bypass ratio (dimensionless)', 'Burner fuel-air ratio (mass ratio, dimensionless)', 'Bleed enthalpy (kJ/kg)', 'Demanded corrected fan speed (rpm)', 'HPT coolant bleed flow (kg/s)', 'LPT coolant bleed flow (kg/s)']
['HPC outlet temperature (°C)']


**Question 5**: <br>
**Compute the number of unique values for the columns with an object and int type.** <br>

*Note: Combine the list with int columns and object columns using the `+` operator*. <br>
*Create a dataframe with the number of unique values and the corresponding variable.*

In [29]:
#Method 1 (Beginner)

# Combine the lists with columns with an object and int type


# Compute the number of unique values in each column
# For each column, we get the number of unique value by using unique
int_obj_columns = int_columns + object_columns
int_obj_uniq = {column: [len(dataset[column].unique())] for column in int_obj_columns}
print(int_obj_uniq)


{'engine_unit_number': [100], 'time_cycles': [362], 'Demanded fan speed (rpm)': [1], 'RUL': [362], 'RUL_class': [2], 'HPC outlet temperature (°C)': [8]}


In [30]:
# Create a dataframe with column names and number of unique values
pd.DataFrame.from_dict(int_obj_uniq)

engine_unit_number  time_cycles  Demanded fan speed (rpm)  RUL  RUL_class  \
0                 100          362                         1  362          2   

   HPC outlet temperature (°C)  
0                            8

In [31]:
#Method 2 (Intermediate) Use nunique
# dataset.select_dtypes(include=['int64','object']).nunique()
dataset[int_obj_columns].nunique()

engine_unit_number             100
time_cycles                    362
Demanded fan speed (rpm)         1
RUL                            362
RUL_class                        2
HPC outlet temperature (°C)      7
dtype: int64

**Which column/variable has over 15 unique values ?**

In [32]:
dataset.loc[:, dataset.nunique() > 15].columns

Index(['engine_unit_number', 'time_cycles', 'setting_1',
       'LPC outlet temperature (°C)', 'LPT outlet temperature (°C)',
       'HPC outlet pressure (bar)', 'Physical fan speed (rpm)',
       'Physical core speed (rpm)', 'HPC outlet static pressure (bar)',
       'Fuel flow / Ps30 (kg/s/bar)', 'Corrected fan speed (rpm)',
       'Corrected core speed (rpm)', 'Bypass ratio (dimensionless)',
       'HPT coolant bleed flow (kg/s)', 'LPT coolant bleed flow (kg/s)',
       'RUL'],
      dtype='object')

**Question 6**: <br>
**Compute the summary statistics of columns with a float type, with pandas' `.describe()` function.** <br>

In [33]:
dataset.select_dtypes(include=['float64']).describe()

setting_1     setting_2  setting_3  Fan inlet temperature (°C)  \
count  20631.000000  20631.000000    20631.0                2.063100e+04   
mean      -0.000009      0.000002      100.0                1.500000e+01   
std        0.002187      0.000293        0.0                2.664600e-14   
min       -0.008700     -0.000600      100.0                1.500000e+01   
25%       -0.001500     -0.000200      100.0                1.500000e+01   
50%        0.000000      0.000000      100.0                1.500000e+01   
75%        0.001500      0.000300      100.0                1.500000e+01   
max        0.008700      0.000600      100.0                1.500000e+01   

       LPC outlet temperature (°C)  LPT outlet temperature (°C)  \
count                 20631.000000                 20631.000000   
mean                     83.894963                   509.590990   
std                       0.277807                     5.000336   
min                      83.077778                   494.766667   
25%                      83.697222                   505.938889   
50%                      83.872222                   509.094444   
75%                      84.072222                   512.713889   
max                      84.922222                   527.677778   

       Fan inlet pressure (bar)  Bypass duct pressure (bar)  \
count              2.063100e+04                20631.000000   
mean               1.008014e+00                    1.489943   
std                2.875547e-13                    0.000096   
min                1.008014e+00                    1.489268   
25%                1.008014e+00                    1.489957   
50%                1.008014e+00                    1.489957   
75%                1.008014e+00                    1.489957   
max                1.008014e+00                    1.489957   

       HPC outlet pressure (bar)  Physical fan speed (rpm)  \
count               20631.000000              20631.000000   
mean                   38.153361               2388.096652   
std                     0.061025                  0.070985   
min                    37.910823               2387.900000   
25%                    38.114908               2388.050000   
50%                    38.158345               2388.090000   
75%                    38.197645               2388.140000   
max                    38.338987               2388.560000   

       Physical core speed (rpm)  \
count               20631.000000   
mean                 9065.242941   
std                    22.082880   
min                  9021.730000   
25%                  9053.100000   
50%                  9060.660000   
75%                  9069.420000   
max                  9244.590000   

       Engine pressure ratio (P50/P2) (dimensionless)  \
count                                    2.063100e+04   
mean                                     1.300000e+00   
std                                      4.660829e-13   
min                                      1.300000e+00   
25%                                      1.300000e+00   
50%                                      1.300000e+00   
75%                                      1.300000e+00   
max                                      1.300000e+00   

       HPC outlet static pressure (bar)  Fuel flow / Ps30 (kg/s/bar)  \
count                      20631.000000                 20631.000000   
mean                           3.277848                    35.950193   
std                            0.018415                     0.050853   
min                            3.230194                    35.762417   
25%                            3.264668                    35.918928   
50%                            3.275699                    35.954780   
75%                            3.288799                    35.987186   
max                            3.346026                    36.085781   

       Corrected fan speed (rpm)  Corrected core speed (rpm)  \
count               20631.000000          

## **4. Analyze missing values**

**Question 7**: <br> **Compute the number of NaN value for every variable/column** <br>

*Note: A NaN value represents a missing value in a cell of the dataframe* <br>
*You can use the `.isna()` function.*

In [34]:
dataset.isna().sum()

engine_unit_number                                      0
time_cycles                                             0
setting_1                                               0
setting_2                                               0
setting_3                                               0
Fan inlet temperature (°C)                              0
LPC outlet temperature (°C)                             0
HPC outlet temperature (°C)                          1866
LPT outlet temperature (°C)                             0
Fan inlet pressure (bar)                                0
Bypass duct pressure (bar)                              0
HPC outlet pressure (bar)                               0
Physical fan speed (rpm)                                0
Physical core speed (rpm)                               0
Engine pressure ratio (P50/P2) (dimensionless)          0
HPC outlet static pressure (bar)                        0
Fuel flow / Ps30 (kg/s/bar)                             0
Corrected fan speed (rpm)                               0
Corrected core speed (rpm)                              0
Bypass ratio (dimensionless)                            0
Burner fuel-air ratio (mass ratio, dimensionless)       0
Bleed enthalpy (kJ/kg)                               1053
Demanded fan speed (rpm)                                0
Demanded corrected fan speed (rpm)                      0
HPT coolant bleed flow (kg/s)                           0
LPT coolant bleed flow (kg/s)                         965
RUL                                                     0
RUL_class                                               0
dtype: int64

** Question 8 Which variables of the dataset has missing values ?**

In [35]:
dataset.loc[:, dataset.isna().sum() > 0]

HPC outlet temperature (°C)  Bleed enthalpy (kJ/kg)  \
0                          Normal                   392.0   
1                          Normal                   392.0   
2                          Normal                   390.0   
3                        Very Low                   392.0   
4                             NaN                   393.0   
...                           ...                     ...   
20626                   Very High                   397.0   
20627              Extremely High                   395.0   
20628                   Very High                   398.0   
20629              Extremely High                   395.0   
20630                   Very High                   396.0   

       LPT coolant bleed flow (kg/s)  
0                          10.622680  
1                          10.624766  
2                          10.588751  
3                          10.602223  
4                          10.616057  
...                              ...  
20626                      10.420604  
20627                      10.504927  
20628                      10.402370  
20629                      10.461654  
20630                      10.456302  

[20631 rows x 3 columns]

**Question 9:** <br>
**Drop the rows of the dataset that have missing values with `.dropna(axis=0)`. <br>**
Don't forget to add `.reset_index(drop=True)` after dropping the NaN values in the dataframe !

In [36]:
dataset.dropna(axis=0).reset_index(drop=True)

engine_unit_number  time_cycles  setting_1  setting_2  setting_3  \
0                       1            1    -0.0007    -0.0004      100.0   
1                       1            2     0.0019    -0.0003      100.0   
2                       1            3    -0.0043     0.0003      100.0   
3                       1            4     0.0007     0.0000      100.0   
4                       1            6    -0.0043    -0.0001      100.0   
...                   ...          ...        ...        ...        ...   
16958                 100          196    -0.0004    -0.0003      100.0   
16959                 100          197    -0.0016    -0.0005      100.0   
16960                 100          198     0.0004     0.0000      100.0   
16961                 100          199    -0.0011     0.0003      100.0   
16962                 100          200    -0.0032    -0.0005      100.0   

       Fan inlet temperature (°C)  LPC outlet temperature (°C)  \
0                            15.0                    83.416667   
1                            15.0                    83.600000   
2                            15.0                    83.711111   
3                            15.0                    83.711111   
4                            15.0                    83.572222   
...                           ...                          ...   
16958                        15.0                    84.344444   
16959                        15.0                    84.372222   
16960                        15.0                    84.305556   
16961                        15.0                    84.200000   
16962                        15.0                    84.544444   

      HPC outlet temperature (°C)  LPT outlet temperature (°C)  \
0                          Normal                   504.961111   
1                          Normal                   506.372222   
2                          Normal                   506.961111   
3                        Very Low                   505.666667   
4                             Low                   503.722222   
...                           ...                          ...   
16958                   Very High                   520.533333   
16959              Extremely High                   523.283333   
16960                   Very High                   520.283333   
16961              Extremely High                   519.366667   
16962                   Very High                   522.483333   

       Fan inlet pressure (bar)  Bypass duct pressure (bar)  \
0                      1.008014                    1.489957   
1                      1.008014                    1.489957   
2                      1.008014                    1.489957   
3                      1.008014                    1.489957   
4                      1.008014                    1.489957   
...                         ...                         ...   
16958                  1.008014                    1.489957   
16959                  1.008014                    1.489957   
16960                  1.008014                    1.489957   
16961                  1.008014                    1.489957   
16962                  1.008014                    1.489957   

       HPC outlet pressure (bar)  Physical fan speed (rpm)  \
0                      38.221777                   2388.06   
1                      38.179719                   2388.04   
2                      38.214882                   2388.08   
3                      38.227982                   2388.11   
4                      38.243150                   2388.02   
...                          ...                       ...   
16958                  38.019760                   2388.19   
16959                  37.980460                   2388.23   
16960                  37.985976                   2388.24   
16961                  37.968049                   2388.25   
16962                  37.975634                   2388.26   

       Physical co

**If you don't want to drop rows, you can replace the missing values in each variable** <br>
Try the following methods only if the variable has a small number of NaN values (less than 10%).
- Replace with the mean or median value for continuous variables (mostly columns with a float dtype)
- Replace with the variable's most frequent value (`.mode()`) or by creating a new category for categorical variables (mostly columns with an int/object dtype)

You can drop the variables with a high number of missing values.

** Re-run the dataset import if you want to do this after having drop the rows**

In [37]:
# Re-run the dataset import if you want to do this after having drop the rows
dataset = pd.read_csv("dataset.csv")
missing_percentages = dataset.isna().sum() / len(dataset) * 100

print("Columns with missing values and their percentage:")
print(missing_percentages[missing_percentages > 0])

for col in missing_percentages[missing_percentages > 0].index:
    if missing_percentages[col] < 10:
        if dataset[col].dtype in ['float64', 'int64']:
            median_val = dataset[col].median()
            dataset[col] = dataset[col].fillna(median_val)
            print(f"Filled missing values in '{col}' with median.")
        elif dataset[col].dtype == 'object':
            mode_val = dataset[col].mode()[0]
            dataset[col] = dataset[col].fillna(mode_val)
            print(f"Filled missing values in '{col}' with mode.")

Columns with missing values and their percentage:
HPC outlet temperature (°C)      9.044642
Bleed enthalpy (kJ/kg)           5.103970
LPT coolant bleed flow (kg/s)    4.677427
dtype: float64
Filled missing values in 'HPC outlet temperature (°C)' with mode.
Filled missing values in 'Bleed enthalpy (kJ/kg)' with median.
Filled missing values in 'LPT coolant bleed flow (kg/s)' with median.


In [38]:
dataset.loc[:,dataset.isna().sum() < 0.1 * len(dataset)]

engine_unit_number  time_cycles  setting_1  setting_2  setting_3  \
0                       1            1    -0.0007    -0.0004      100.0   
1                       1            2     0.0019    -0.0003      100.0   
2                       1            3    -0.0043     0.0003      100.0   
3                       1            4     0.0007     0.0000      100.0   
4                       1            5    -0.0019    -0.0002      100.0   
...                   ...          ...        ...        ...        ...   
20626                 100          196    -0.0004    -0.0003      100.0   
20627                 100          197    -0.0016    -0.0005      100.0   
20628                 100          198     0.0004     0.0000      100.0   
20629                 100          199    -0.0011     0.0003      100.0   
20630                 100          200    -0.0032    -0.0005      100.0   

       Fan inlet temperature (°C)  LPC outlet temperature (°C)  \
0                            15.0                    83.416667   
1                            15.0                    83.600000   
2                            15.0                    83.711111   
3                            15.0                    83.711111   
4                            15.0                    83.722222   
...                           ...                          ...   
20626                        15.0                    84.344444   
20627                        15.0                    84.372222   
20628                        15.0                    84.305556   
20629                        15.0                    84.200000   
20630                        15.0                    84.544444   

      HPC outlet temperature (°C)  LPT outlet temperature (°C)  \
0                          Normal                   504.961111   
1                          Normal                   506.372222   
2                          Normal                   506.961111   
3                        Very Low                   505.666667   
4                          Normal                   508.083333   
...                           ...                          ...   
20626                   Very High                   520.533333   
20627              Extremely High                   523.283333   
20628                   Very High                   520.283333   
20629              Extremely High                   519.366667   
20630                   Very High                   522.483333   

       Fan inlet pressure (bar)  Bypass duct pressure (bar)  \
0                      1.008014                    1.489957   
1                      1.008014                    1.489957   
2                      1.008014                    1.489957   
3                      1.008014                    1.489957   
4                      1.008014                    1.489957   
...                         ...                         ...   
20626                  1.008014                    1.489957   
20627                  1.008014                    1.489957   
20628                  1.008014                    1.489957   
20629                  1.008014                    1.489957   
20630                  1.008014                    1.489957   

       HPC outlet pressure (bar)  Physical fan speed (rpm)  \
0                      38.221777                   2388.06   
1                      38.179719                   2388.04   
2                      38.214882                   2388.08   
3                      38.227982                   2388.11   
4                      38.196955                   2388.06   
...                          ...                       ...   
20626                  38.019760                   2388.19   
20627                  37.980460                   2388.23   
20628                  37.985976                   2388.24   
20629                  37.968049                   2388.25   
20630                  37.975634                   2388.26   

       Physical co

At this step, the dataset shouldn't have any missing values (you can check with `.isna().sum().sum()`)


In [39]:
dataset.isna().sum().sum()

np.int64(0)

In [40]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20631 entries, 0 to 20630
Data columns (total 28 columns):
 #   Column                                             Non-Null Count  Dtype  
---  ------                                             --------------  -----  
 0   engine_unit_number                                 20631 non-null  int64  
 1   time_cycles                                        20631 non-null  int64  
 2   setting_1                                          20631 non-null  float64
 3   setting_2                                          20631 non-null  float64
 4   setting_3                                          20631 non-null  float64
 5   Fan inlet temperature (°C)                         20631 non-null  float64
 6   LPC outlet temperature (°C)                        20631 non-null  float64
 7   HPC outlet temperature (°C)                        20631 non-null  object 
 8   LPT outlet temperature (°C)                        20631 non-null  float64
 9   Fan in

**Question 11**: <br>
**Save the cleaned dataframe as a csv file called `dataset_train_clean.csv` using pandas' `.to_csv()` function.** <br>
*Note: Make sure to add `index=False` to the `.to_csv()` function or else the index of the dataframe will be saved too.*

In [41]:
dataset.to_csv("dataset_clean.csv",index=False)